lib/Conversion/StandardToHandshake/MaximizeSSA.cpp

```diff
@@ -0,0 +1,230 @@
+  for (uint idx = 0, numSucc = successors.size(); idx < numSucc; ++idx) {
+    auto *succBlock = successors[idx];
+    if (succBlock == block) {
+      branchOp.getSuccessorOperands(idx).append(value);
+    }
+  }
+
+  return success();
+}
+
+namespace circt {
+
+bool isRegionSSAMaximized(Region &region) {
+
+  // Check whether all operands used within each block are also defined within
+  // the same block
+  for (auto &block : region.getBlocks())
+    for (auto &op : block.getOperations())
+      for (auto operand : op.getOperands())
+        if (getDefiningBlock(operand) != &block)
+          return false;
+
+  return true;
+}
+
+LogicalResult maximizeSSA(Value value, PatternRewriter &rewriter) {
+
+  // Identify the basic block in which the value is defined
+  Block *defBlock = getDefiningBlock(value);
+
+  // Identify all basic blocks in which the value is used (excluding the
+  // value-defining block)
+  DenseSet<Block *> blocksUsing;
+  for (auto &use : value.getUses()) {
+    auto *block = use.getOwner()->getBlock();
+    if (block != defBlock)
+      blocksUsing.insert(block);
+  }
+
+  // Prepare a stack to store the list of basic blocks that must be modified for
+  // the value to be in maximum SSA form. At all points, blocksUsing is a
+  // non-strict superset of the elements contained in blocksToVisit
+  std::stack<Block *> blocksToVisit;
+  for (auto *block : blocksUsing) {
+    blocksToVisit.push(block);
+  }
+
+  // Backtrack from all blocks using the value to the value-defining basic
+  // block, adding a new block argument for the value along the way. Keep
+  // track of which blocks have already been modified to avoid visiting a block
+  // more than once while backtracking (possible due to branching control flow)
+  DenseMap<Block *, BlockArgument> blockToArg;
+  while (!blocksToVisit.empty()) {
+    // Retrieve the basic block at the top of the stack, remove it from the
+    // stack, and add it to the list of visited blocks
+    auto *block = blocksToVisit.top();
+    blocksToVisit.pop();
+
+    // Add an argument to the block to hold the value
+    blockToArg[block] =
+        block->addArgument(value.getType(), rewriter.getUnknownLoc());
+
+    // In all block predecessors, modify the terminator branching instruction to
+    // also pass the value to the block
+    for (auto *predBlock : block->getPredecessors()) {
+
+      // Modify the terminator instruction
+      if (failed(addArgToTerminator(block, predBlock, value)))
+        return failure();
+
+      // Now the predecessor block is using the value, so we must also make sure
+      // to visit it
+      if (predBlock != defBlock) {
+        auto [_, blockNewlyUsing] = blocksUsing.insert(predBlock);
+        if (blockNewlyUsing) {
+          blocksToVisit.push(predBlock);
+        }
+      }
+    }
+  }
+
+  // Replace all uses of the value with the newly added block arguments
+  SmallVector<Operation *> users;
+  for (auto &use : value.getUses()) {
```

Comment:
If I were to use `llvm::copy_if` here it seems that I would also need to do a `llvm::transform` first since I want to get each use's owner. This means I would have to create a temporary array of all owners and then filter it using `llvm::copy_if`, which I don't currently need to do. Or did I misunderstand your comment?

Here is what I mean:
```cpp
SmallVector<Operation *> users, usersFiltered;
llvm::transform(value.getUses(), users, [](auto use) { return use.getOwner(); });
llvm::copy_if(users, usersFiltered, [&](auto user) { return user->getBlock() != defBlock; });
for (auto *user : users)
  user->replaceUsesOfWith(value, blockToArg[user->getBlock()]);
```

Reply:
Sorry for the oversight, I might have been a bit tired. In this case, using a foreach style loop as you did is fine.